### ISSUE TITLE
sql: slow APPLY JOIN query with virtual tables

### ISSUE THREAD

**Issue by otan** [cockroachdb/cockroach#77877]
Consider the following query:

```
SELECT
    EXISTS(
        SELECT
            1
        FROM
            pg_attribute AS at
            LEFT JOIN pg_init_privs AS pip ON
                    c.oid = pip.objoid
                    AND pip.classoid = 'pg_class'::REGCLASS
                    AND pip.objsubid = at.attnum
        WHERE
            at.attrelid = c.oid
            AND (
                    (
                        SELECT
                            pg_catalog.array_agg(
                                acl ORDER BY row_n
                            )
                        FROM
                            (
                                SELECT
                                    acl, row_n
                                FROM
                                    pg_catalog.unnest(
                                            at.attacl
                                    ) WITH ORDINALITY
                                        AS perm (acl, row_n)
                                WHERE
                                    NOT
                                        EXISTS(
                                            SELECT
                                                1
                                            FROM
                                                pg_catalog.unnest(
                                                        pip.initprivs
                                                )
                                                    AS init
                                                        (
                                                            init_acl
                                                        )
                                            WHERE
                                                acl
                                                = init_acl
                                        )
                            )
                                AS foo
                    ) IS NOT NULL
                    OR (
                            SELECT
                                pg_catalog.array_agg(
                                    acl ORDER BY row_n
                                )
                            FROM
                                (
                                    SELECT
                                        acl, row_n
                                    FROM
                                        pg_catalog.unnest(
                                                pip.initprivs
                                        ) WITH ORDINALITY
                                            AS initp
                                                (acl, row_n)
                                    WHERE
                                        NOT
                                            EXISTS(
                                                SELECT
                                                    1
                                                FROM
                                                    pg_catalog.unnest(
                                                            at.attacl
                                                    )
                                                        AS permp
                                                            (
                                                                orig_acl
                                                            )
                                                WHERE
                                                    acl
                                                    = orig_acl
                                            )
                                )
                                    AS foo
                        ) IS NOT NULL
                    OR NULL IS NOT NULL
                    OR NULL IS NOT NULL
                )
    )
        AS changed_acl
FROM
    pg_class AS c
    LEFT JOIN pg_class AS tc ON
            c.reltoastrelid = tc.oid AND c.relkind != 'p'
    LEFT JOIN pg_am AS am ON c.relam = am.oid
    LEFT JOIN pg_init_privs AS pip ON
            c.oid = pip.objoid
            AND pip.classoid = 'pg_class'::REGCLASS
            AND pip.objsubid = 0
WHERE
    c.relkind IN ('r', 'S', 'v', 'c', 'm', 'f', 'p')
ORDER BY
    c.oid;

```

This query takes minutes to run, despite no rows in `pg_init_privs`.

**Comment by msirek** [comment]
TLDR: Apply join can be slow because it plans and runs the whole subquery for each left row.
Correlated joins that use apply join could be improved by either:
1. Avoiding the apply join by pushing the outer table into the subquery while retaining the outer referenced table, then semi-joining on rowid.
2. Using placeholders for the left row values of the apply join and building a reusable memo to avoid plan rebuilding costs.

Here's a CPU profile snippet showing apply join spending most of the time in the execbuilder:
![image](https://user-images.githubusercontent.com/3999975/161355320-fa270831-d915-45a6-8c19-1c487ebb3269.png)

A simplified problem query:
```
SELECT COUNT(*)
FROM
    pg_class AS c
WHERE
    EXISTS(
        SELECT
            1
        FROM
            pg_attribute AS at
            LEFT JOIN pg_init_privs AS pip ON
                    c.oid = pip.objoid
                    AND pip.classoid = 'pg_class'::REGCLASS
                    AND pip.objsubid = at.attnum
        WHERE
            at.attrelid = c.oid
    )
;
  count
---------
    297
(1 row)

Time: 7.578s

                       info
---------------------------------------------------
  scalar-group-by
   ├── semi-join-apply
   │    ├── scan pg_class [as=c]
   │    ├── left-join (hash)
   │    │    ├── scan pg_attribute [as=at]
   │    │    ├── select
   │    │    │    ├── scan pg_init_privs [as=pip]
   │    │    │    └── filters
   │    │    │         └── classoid = pg_class
   │    │    └── filters
   │    │         ├── oid = objoid
   │    │         └── objsubid = attnum
   │    └── filters
   │         └── attrelid = oid
   └── aggregations
        └── count-rows
```
Manually rewritten query:
```
CREATE TABLE mws.pg_class2 AS SELECT * FROM pg_class;
delete from  mws.pg_class2;
insert into mws.pg_class2 select * from pg_class;
SELECT COUNT(*) from mws.pg_class2 abc WHERE rowid IN
        (SELECT
            c.rowid
        FROM
            mws.pg_class2 AS c INNER JOIN
            pg_attribute AS at
         ON at.attrelid = c.oid

            LEFT JOIN pg_init_privs AS pip ON
                    c.oid = pip.objoid
                    AND pip.classoid = 'pg_class'::REGCLASS
                    AND pip.objsubid = at.attnum
);
  count
---------
    297
(1 row)

Time: 58ms
                      info
-------------------------------------------------
  scalar-group-by
   ├── inner-join (hash)
   │    ├── scan pg_class2 [as=abc]
   │    ├── distinct-on
   │    │    └── inner-join (hash)
   │    │         ├── scan pg_class2 [as=c]
   │    │         ├── scan pg_attribute [as=at]
   │    │         └── filters
   │    │              └── attrelid = c.oid
   │    └── filters
   │         └── abc.rowid = c.rowid
   └── aggregations
        └── count-rows
```
The problem with the query rewrite is that the `pg_class` is a virtual table, which doesn't appear to have a `rowid` column (at least not exposed via SQL).  I tried using a CTE to assign dummy rowids, which appears to work, so this is still a viable option for virtual tables even if there's no `rowid` column:
```
> WITH c as (SELECT pg_class.*, unique_rowid() rid from pg_class)
SELECT count(*) FROM c WHERE c.rid IN
        (SELECT
            pushed_c.rid
        FROM
            c pushed_c INNER JOIN
            pg_attribute AS at
            ON
              at.attrelid = pushed_c.oid
            LEFT JOIN pg_init_privs AS pip ON
                    pushed_c.oid = pip.objoid
                    AND pip.classoid = 'pg_class'::REGCLASS
                    AND pip.objsubid = at.attnum
       );
  count
---------
    297
(1 row)

Time: 76ms
                         info
------------------------------------------------------
  with &1 (c)
   ├── project
   │    ├── scan pg_class
   │    └── projections
   │         └── unique_rowid()
   └── project
        └── inner-join (hash)
             ├── with-scan &1 (c)
             ├── distinct-on
             │    └── inner-join (hash)
             │         ├── with-scan &1 (c)
             │         ├── scan pg_attribute [as=at]
             │         └── filters
             │              └── attrelid = oid
             └── filters
                  └── rid = rid
``` 
I'm not sure which solution is better, but creating the plan as a prepared statement (using placeholders) seems like it could be applied in more cases, since there are probably many semijoin/antijoin predicates and combinations involving outer join for which pushing the outer table would be illegal.

**Comment by msirek** [comment]
branch:  WithClauseDecorrelate

**Comment by mgartner** [comment]
Running on v23.1.2:

The original query in the issue description now take ~620ms instead of minutes. 🎉 

This query is much faster now, taking only ~100ms (@msirek reported it taking ~7.5s above):

```
SELECT COUNT(*)
FROM
    pg_class AS c
WHERE
    EXISTS(
        SELECT
            1
        FROM
            pg_attribute AS at
            LEFT JOIN pg_init_privs AS pip ON
                    c.oid = pip.objoid
                    AND pip.classoid = 'pg_class'::REGCLASS
                    AND pip.objsubid = at.attnum
        WHERE
            at.attrelid = c.oid
    )
;

                                 info
----------------------------------------------------------------------
  scalar-group-by
   ├── inner-join (merge)
   │    ├── scan pg_class@pg_class_oid_idx [as=c]
   │    ├── distinct-on
   │    │    └── scan pg_attribute@pg_attribute_attrelid_idx [as=at]
   │    └── filters (true)
   └── aggregations
        └── count-rows
```
